File: converter/src/services/converter.rs
use bytes::Bytes;
use image::{DynamicImage, ImageOutputFormat, imageops::FilterType};
use std::io::Cursor;

use crate::error::ConverterError;

#[derive(Debug, Clone, Copy, Default)]
pub enum OutputFormat {
    Png,
    #[default]
    WebP,
}

#[derive(Debug, Clone, Copy, Default)]
pub enum FitMode {
    #[default]
    Cover,
    Contain,
    Fill,
}

impl FitMode {
    pub fn from_str(s: &str) -> Option<Self> {
        match s.to_lowercase().as_str() {
            "cover" => Some(FitMode::Cover),
            "contain" => Some(FitMode::Contain),
            "fill" => Some(FitMode::Fill),
            _ => None,
        }
    }
}

pub struct ConversionRequest {
    pub data: Bytes,
    pub width: Option<u32>,
    pub height: Option<u32>,
    pub output_format: OutputFormat,
    pub quality: u8,
    pub fit_mode: FitMode,
}

pub struct ConversionResult {
    pub data: Bytes,
    pub content_type: &'static str,
    pub original_width: u32,
    pub original_height: u32,
    pub output_width: u32,
    pub output_height: u32,
}

pub fn convert(request: ConversionRequest) -> Result<ConversionResult, ConverterError> {
    // Load image from bytes
    let img = image::load_from_memory(&request.data)?;
    let original_width = img.width();
    let original_height = img.height();

    // Resize if dimensions are specified
    let resized = resize_image(&img, request.width, request.height, request.fit_mode);
    let output_width = resized.width();
    let output_height = resized.height();

    // Encode to output format
    let (data, content_type) = encode_image(&resized, request.output_format, request.quality)?;

    Ok(ConversionResult {
        data,
        content_type,
        original_width,
        original_height,
        output_width,
        output_height,
    })
}

fn resize_image(
    img: &DynamicImage,
    width: Option<u32>,
    height: Option<u32>,
    fit_mode: FitMode,
) -> DynamicImage {
    match (width, height) {
        (Some(w), Some(h)) => {
            match fit_mode {
                FitMode::Cover => img.resize_to_fill(w, h, FilterType::Lanczos3),
                FitMode::Contain => img.resize(w, h, FilterType::Lanczos3),
                FitMode::Fill => img.resize_exact(w, h, FilterType::Lanczos3),
            }
        }
        (Some(w), None) => {
            // Maintain aspect ratio based on width
            let ratio = w as f64 / img.width() as f64;
            let new_height = (img.height() as f64 * ratio) as u32;
            img.resize_exact(w, new_height, FilterType::Lanczos3)
        }
        (None, Some(h)) => {
            // Maintain aspect ratio based on height
            let ratio = h as f64 / img.height() as f64;
            let new_width = (img.width() as f64 * ratio) as u32;
            img.resize_exact(new_width, h, FilterType::Lanczos3)
        }
        (None, None) => img.clone(),
    }
}

fn encode_image(
    img: &DynamicImage,
    format: OutputFormat,
    quality: u8,
) -> Result<(Bytes, &'static str), ConverterError> {
    match format {
        OutputFormat::Png => {
            let mut buffer = Vec::new();
            img.write_to(&mut Cursor::new(&mut buffer), ImageOutputFormat::Png)?;
            Ok((Bytes::from(buffer), "image/png"))
        }
        OutputFormat::WebP => {
            let rgba = img.to_rgba8();
            let encoder = webp::Encoder::from_rgba(&rgba, img.width(), img.height());
            let webp_data = encoder.encode(quality as f32);
            Ok((Bytes::from(webp_data.to_vec()), "image/webp"))
        }
    }
}

File: converter/converter.go
package main

import (
	"bytes"
	"fmt"
	"image"
	"image/jpeg"
	_ "image/gif"
	_ "image/png"

	"github.com/chai2010/webp"
	"github.com/disintegration/imaging"
	"github.com/gen2brain/avif"
)

type OutputFormat int

const (
	OutputFormatPNG OutputFormat = iota
	OutputFormatAVIF
	OutputFormatWebP
	OutputFormatJPEG
)

type FitMode int

const (
	FitModeCover FitMode = iota
	FitModeContain
	FitModeFill
)

func (f FitMode) String() string {
	switch f {
	case FitModeCover:
		return "cover"
	case FitModeContain:
		return "contain"
	case FitModeFill:
		return "fill"
	default:
		return "cover"
	}
}

func ParseFitMode(s string) FitMode {
	switch s {
	case "cover":
		return FitModeCover
	case "contain":
		return FitModeContain
	case "fill":
		return FitModeFill
	default:
		return FitModeCover
	}
}

type ConversionRequest struct {
	Data         []byte
	Width        *uint32
	Height       *uint32
	OutputFormat OutputFormat
	Quality      int
	FitMode      FitMode
}

type ConversionResult struct {
	Data           []byte
	ContentType    string
	OriginalWidth  uint32
	OriginalHeight uint32
	OutputWidth    uint32
	OutputHeight   uint32
}

func Convert(req *ConversionRequest) (*ConversionResult, error) {
	// Decode image
	img, _, err := image.Decode(bytes.NewReader(req.Data))
	if err != nil {
		return nil, fmt.Errorf("failed to decode image: %w", err)
	}

	bounds := img.Bounds()
	originalWidth := uint32(bounds.Dx())
	originalHeight := uint32(bounds.Dy())

	// Resize if dimensions are specified
	resized := resizeImage(img, req.Width, req.Height, req.FitMode)
	resizedBounds := resized.Bounds()
	outputWidth := uint32(resizedBounds.Dx())
	outputHeight := uint32(resizedBounds.Dy())

	// Encode to output format
	data, contentType, err := encodeImage(resized, req.OutputFormat, req.Quality)
	if err != nil {
		return nil, err
	}

	return &ConversionResult{
		Data:           data,
		ContentType:    contentType,
		OriginalWidth:  originalWidth,
		OriginalHeight: originalHeight,
		OutputWidth:    outputWidth,
		OutputHeight:   outputHeight,
	}, nil
}

func resizeImage(img image.Image, width, height *uint32, fitMode FitMode) image.Image {
	if width == nil && height == nil {
		return img
	}

	bounds := img.Bounds()
	origWidth := uint32(bounds.Dx())
	origHeight := uint32(bounds.Dy())

	var newWidth, newHeight int

	if width != nil && height != nil {
		// Both dimensions specified
		switch fitMode {
		case FitModeCover:
			// Resize to fill, cropping if necessary
			return imaging.Fill(img, int(*width), int(*height), imaging.Center, imaging.Lanczos)
		case FitModeContain:
			// Resize to fit within bounds, maintaining aspect ratio
			return imaging.Fit(img, int(*width), int(*height), imaging.Lanczos)
		case FitModeFill:
			// Resize exactly, ignoring aspect ratio
			return imaging.Resize(img, int(*width), int(*height), imaging.Lanczos)
		}
	} else if width != nil {
		// Only width specified - maintain aspect ratio
		ratio := float64(*width) / float64(origWidth)
		newWidth = int(*width)
		newHeight = int(float64(origHeight) * ratio)
	} else {
		// Only height specified - maintain aspect ratio
		ratio := float64(*height) / float64(origHeight)
		newHeight = int(*height)
		newWidth = int(float64(origWidth) * ratio)
	}

	return imaging.Resize(img, newWidth, newHeight, imaging.Lanczos)
}

func encodeImage(img image.Image, format OutputFormat, quality int) ([]byte, string, error) {
	var buf bytes.Buffer

	switch format {
	case OutputFormatPNG:
		if err := imaging.Encode(&buf, img, imaging.PNG); err != nil {
			return nil, "", fmt.Errorf("failed to encode PNG: %w", err)
		}
		return buf.Bytes(), "image/png", nil

	case OutputFormatAVIF:
		// gen2brain/avif Quality: 0-100, higher is better (100 = lossless)
		// Use YCbCrSubsampleRatio444 to preserve color accuracy
		opts := avif.Options{
			Quality:           quality,
			QualityAlpha:      quality,
			Speed:             6,
			ChromaSubsampling: image.YCbCrSubsampleRatio444, // Full chroma for better color
		}
		if err := avif.Encode(&buf, img, opts); err != nil {
			return nil, "", fmt.Errorf("failed to encode AVIF: %w", err)
		}
		return buf.Bytes(), "image/avif", nil

	case OutputFormatWebP:
		if err := webp.Encode(&buf, img, &webp.Options{Quality: float32(quality)}); err != nil {
			return nil, "", fmt.Errorf("failed to encode WebP: %w", err)
		}
		return buf.Bytes(), "image/webp", nil

	case OutputFormatJPEG:
		if err := jpeg.Encode(&buf, img, &jpeg.Options{Quality: quality}); err != nil {
			return nil, "", fmt.Errorf("failed to encode JPEG: %w", err)
		}
		return buf.Bytes(), "image/jpeg", nil

	default:
		return nil, "", fmt.Errorf("unsupported output format")
	}
}
